sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.c

```diff
@@ -0,0 +1,485 @@
+/*
+ * Copyright (C) 2017 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @author  Martine Lenders <m.lenders@fu-berlin.de>
+ */
+
+#include "xtimer.h"
+#include "net/gnrc/ndp2.h"
+#include "net/gnrc/ipv6/nib.h"
+
+#include "_nib-arsm.h"
+#include "_nib-6lr.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+static char addr_str[IPV6_ADDR_MAX_STR_LEN];
+#endif
+
+/**
+ * @brief   Determines supposed link-layer address from interface and option
+ *          length
+ *
+ * @param[in] netif A network interface.
+ * @param[in] opt   A SL2AO or TL2AO.
+ *
+ * @return  The length of the L2 address carried in @p opt.
+ */
+static inline unsigned _get_l2addr_len(gnrc_ipv6_netif_t *netif,
+                                       const ndp_opt_t *opt);
+
+void _snd_ns(const ipv6_addr_t *tgt, gnrc_ipv6_netif_t *netif,
+             const ipv6_addr_t *src, const ipv6_addr_t *dst)
+{
+    gnrc_pktsnip_t *ext_opt = NULL;
+
+    gnrc_ndp2_nbr_sol_send(tgt, netif, src, dst, ext_opt);
+}
+
+void _snd_uc_ns(_nib_onl_entry_t *nbr, bool reset)
+{
+    gnrc_ipv6_netif_t *netif = gnrc_ipv6_netif_get(_nib_onl_get_if(nbr));
+    _nib_iface_t *iface = _nib_iface_get(_nib_onl_get_if(nbr));
+
+    DEBUG("unicast to %s (retrans. timer = %ums)\n",
+          ipv6_addr_to_str(addr_str, &nbr->ipv6, sizeof(addr_str)),
+          (unsigned)iface->retrans_time);
+    assert((netif != NULL) && (iface != NULL));
+#if GNRC_IPV6_NIB_CONF_ARSM
```

Comment:
Because sending Neighbor Solicitations is (originally) related to address resolution, but also required for other stuff (e.g. 6Lo address registration and SLAAC). Welcome to NDP hell ;-).

Reply:
(the ARSM-only part starts in l172)